algorithms/src/Kokkos_Random.hpp

```diff
@@ -903,24 +903,45 @@ class Random_XorShift64_Pool {
 #else
   Random_XorShift64_Pool()   = default;
 #endif
+
   Random_XorShift64_Pool(uint64_t seed) {
-    num_states_ = 0;
+    init(execution_space(), seed, execution_space().concurrency());
+    execution_space().fence();
+  }
+
+  Random_XorShift64_Pool(uint64_t seed, uint64_t num_states) {
+    init(execution_space(), seed, num_states);
+    execution_space().fence();
+  }
 
-    init(seed, execution_space().concurrency());
+  Random_XorShift64_Pool(const execution_space& exec, uint64_t seed) {
+    init(exec, seed, exec.concurrency());
   }
 
-  void init(uint64_t seed, int num_states) {
+  Random_XorShift64_Pool(const execution_space& exec, uint64_t seed,
+                         uint64_t num_states) {
+    init(exec, seed, num_states);
+  }
+
+  void init(uint64_t seed, uint64_t num_states) {
+    init(execution_space(), seed, num_states);
+  }
+
+ private:
+  void init(execution_space const& exec, uint64_t seed, uint64_t num_states) {
```

Comment:
should we call this `init_impl` ? it might give easier to read/reason compiler errors for users if the name is different

Reply:
Why don't we expose this overload to the user?